piq/vsi.py

```diff
@@ -0,0 +1,285 @@
+r"""Implemetation of Visual Saliency-induced Index
+Code is based on MATLAB version for computations in pixel domain
+https://sse.tongji.edu.cn/linzhang/IQA/VSI/VSI.htm
+
+References:
+    https://ieeexplore.ieee.org/document/6873260
+"""
+
+import torch
+from torch.nn.modules.loss import _Loss
+from torch.nn.functional import conv2d, avg_pool2d, interpolate
+from .utils import _validate_input, _adjust_dimensions
+
+
+def vsi(prediction: torch.Tensor, target: torch.Tensor, reduction: str = 'mean', data_range: float = 1.,
+        c1: float = 1.27, c2: float = 386., c3: float = 130., alpha: float = 0.4, beta: float = 0.02,
+        omega_0: float = 0.021, sigma_f: float = 1.34, sigma_d: float = 145., sigma_c: float = 0.001) -> torch.Tensor:
+    r"""Compute Visual Saliency-induced Index for a batch of images.
+
+        Both inputs supposed to have RGB order.
+        Args:
+            prediction: Batch of predicted images with shape (batch_size x channels x H x W)
+            target: Batch of target images with shape  (batch_size x channels x H x W)
+            reduction: Reduction over samples in batch: "mean"|"sum"|"none"
+            data_range: Value range of input images (usually 1.0 or 255). Default: 1.0
+            c1: coefficient to calculate saliency component of VSI
+            c2: coefficient to calculate gradient component of VSI
+            c3: coefficient to calculate color component of VSI
+            alpha: power for gradient component of VSI
+            beta: power for color component of VSI
+            omega_0: coefficient to get log Gabor filter at SDSP
+            sigma_f: coefficient to get log Gabor filter at SDSP
+            sigma_d: coefficient to get SDSP
+            sigma_c: coefficient to get SDSP
+
+        Returns:
+            VSI: Index of similarity between two images. Usually in [0, 1] interval.
+        Note:
+            See https://ieeexplore.ieee.org/document/6873260 for details.
+        """
+
+    _validate_input(input_tensors=(prediction, target), allow_5d=False)
+    prediction, target = _adjust_dimensions(input_tensors=(prediction, target))
+    assert prediction.size(-3) == 3 and target.size(-3) == 3, 'Expected RGB images, got images with ' \
```

Comment:
Is it possible to adjust the implementation to support grayscale images?

Reply:
The method requires all 3 channels, which are processed differently. One could convert greyscale image to RGB using the best suitable approach for their task.
We can introduce greyscale image processing copying the image three times, but it could create wrong user expectation from the measure. It is a workaround, which is not described in the paper. I would suggest to keep supporting RGB only images.
@zakajd @snk4tr Any other thoughts?

Reply:
I would propose to add support for greyscale images and mention this limitation in docstring.